flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBaseTest.java

```diff
@@ -204,23 +206,113 @@ public void go() throws Exception {
 			Thread.sleep(5);
 		}
 
-		Assert.assertEquals(1, casSinkFunc.getNumOfPendingRecords());
+		Assert.assertEquals(1, casSinkFunc.getAcquiredPermits());
 		completableFuture.complete(null);
-		Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+		Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
 	}
 
-	private static class TestCassandraSink extends CassandraSinkBase<String, ResultSet> {
+	@Test(timeout = 5000)
+	public void testAcquireOnInvoke() throws Exception {
+		TestCassandraSink casSinkFunc = new TestCassandraSink();
+		casSinkFunc.setMaxConcurrentRequests(1, 5, TimeUnit.MILLISECONDS);
+
+		casSinkFunc.open(new Configuration());
+
+		CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+
+		casSinkFunc.setResultSetFuture(ResultSetFutures.fromCompletableFuture(completableFuture));
+
+		Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+		Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+
+		casSinkFunc.invoke("hello");
+
+		Assert.assertEquals(0, casSinkFunc.getAvailablePermits());
+		Assert.assertEquals(1, casSinkFunc.getAcquiredPermits());
+
+		completableFuture.complete(null);
+
+		casSinkFunc.close();
+	}
+
+	@Test(timeout = 5000)
+	public void testReleaseOnSuccess() throws Exception {
+		TestCassandraSink casSinkFunc = new TestCassandraSink();
+		casSinkFunc.setMaxConcurrentRequests(1, 5, TimeUnit.MILLISECONDS);
+
+		casSinkFunc.open(new Configuration());
+
+		CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+
+		casSinkFunc.setResultSetFuture(ResultSetFutures.fromCompletableFuture(completableFuture));
+
+		Assert.assertEquals(1, casSinkFunc.getAvailablePermits());
+		Assert.assertEquals(0, casSinkFunc.getAcquiredPermits());
+
+		casSinkFunc.invoke("hello");
```

Comment:
The `CompletableFuture`'s could be added to the list inside `invoke`, queried from `TestCassandraSink` and completed later in any order. This way we could check multiple pending invokes.

Reply:
I added a `Queue` to the `TestCassandraSink`.